### PR TITLE
2fa sunday

### PR DESCRIPTION
**Pull request by kezuk23**
The pull request from the work from last sprint. It's been refactored into its own class and hopefully the implementation is correct. Sorry if it's a little bit messy!

### CODE REVIEW

File: src/main/java/server/InputProcessor.java [hunk 1]
```diff
@@ -287,4 +288,29 @@ private static boolean isValid(String input, String regex){
...
+                inputCode = in.readLine();
+                // Check if the provided input is valid by comparing its format with the
+                // relevant regex
+                // Upon matching, break and return the given input
+                if (isValid(inputCode, InfoRegexMap.get(key))) {
+                    break;
+                } else {
+                    // Display a message if the input is invalid. Loop again to take a new input
+                    out.println("Invalid entry. This field accepts " + key + " only. \nTry again:");
+                    continue;
```
Comment (m-c-hill) [inline src/main/java/server/InputProcessor.java:306]: This continue can be removed as the code will continue once it reaches this point regardless

File: src/main/java/server/TwoFactorAuth.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    private static String expectedCode = "null";
+    private static boolean authentication;
+
+    // generate random 4 digit code - store as String 'expectedCode'
+    public static String generateCode() {
+
+        Random random = new Random();
+        String expectedCode = String.format("%04d", random.nextInt(10000));
+        return expectedCode;
```
Comment (m-c-hill) [inline src/main/java/server/TwoFactorAuth.java:17]: Good use of random number generation!

File: src/main/java/server/TwoFactorAuth.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+        BufferedReader in;
+        PrintWriter out;
+
+        // authorisation false until correct input
+        boolean authentication = false;
+
+        // askforcodeinput
+        System.out.println("Please enter the four digit code:");
+        String inputCode = InputProcessor.takeCodeInput("four digit code", in, out);
```
Comment (m-c-hill) [inline src/main/java/server/TwoFactorAuth.java:30]: You'll need to initialise the BufferedReader and Printwriter objects as you're currently passing in uninitialised objects.

File: src/main/java/server/InputProcessor.java [hunk 4]
```diff
@@ -22,7 +22,8 @@
         "valid email addresses", "^(.+)@(.+)[\\.]{1}(\\D+)$",
         "valid dates", "^(0[1-9]|[12][0-9]|3[01])[-/. ]?([0][1-9]|[1][012])[-/. ]?(19|20)\\d\\d$",
         "valid phone numbers", "^0[\\d]{7,12}$",
-        "valid postcodes/zipcodes", "^[a-zA-Z0-9 ]{3,10}$"
+        "valid postcodes/zipcodes", "^[a-zA-Z0-9 ]{3,10}$",
+        "four digit codes", "^[0-9]{4}$"
```
Comment (m-c-hill) [inline src/main/java/server/InputProcessor.java:26]: Good additions to the regex list!

File: src/main/java/server/TwoFactorAuth.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        Sms.sendText(generateCode());
+
+        System.out.println("Your one-time passcode has been sent to your phone");
+
+        codeInput();
+
+        return authentication;
+    }
+
+}
```
Comment (m-c-hill) [inline src/main/java/server/TwoFactorAuth.java:51]: Very clean and easy to read!

**Review by m-c-hill — CHANGES_REQUESTED** [review]
2FA logic is very well thought out and is a great feature to add to the NewBank app, well done! Code is clearl and easy to read.

The one change that needs to be made for this to work is to initialise your BufferedReader and PrintWriter in the codeInput method.

**Review by Ecenursahin — COMMENTED** [review]
Simon, this is a very clean implementation and a great addition to our bank system. I agree with Matt's suggestion. Thank you for working on this element!

File: src/main/java/server/TwoFactorAuth.java (hunk 3 above)
Comment (kezuk23) [reply src/main/java/server/TwoFactorAuth.java:32]: Added

**Review by AllenNinous — APPROVED** [review]
Nicely done Simon. It's also nice to see that this functionality is encapsulated in a class of its own. Great job.

**Review by kezuk23 — COMMENTED** [review]
Thanks for that Ben, I can't approve this so we just need somebody else to approve before merging! @Ecenursahin @AllenNinous are you able to approve and merge?

**Review by m-c-hill — APPROVED** [review]

**Pull request merged**